clients/da-vinci-client/src/main/java/com/linkedin/davinci/storage/StorageService.java

```diff
@@ -371,6 +375,40 @@ public synchronized AbstractStorageEngine openStore(
     return engine;
   }
 
+  public synchronized void checkWhetherStoragePartitionsShouldBeKeptOrNot(SafeHelixManager manager) {
+    if (manager == null) {
+      return;
+    }
+    for (AbstractStorageEngine storageEngine: getStorageEngineRepository().getAllLocalStorageEngines()) {
+      String storeName = storageEngine.getStoreVersionName();
+      Set<Integer> storageEnginePartitionIds = new HashSet<>(storageEngine.getPartitionIds());
+      String instanceHostName = manager.getInstanceName();
+      PropertyKey.Builder propertyKeyBuilder =
+          new PropertyKey.Builder(configLoader.getVeniceClusterConfig().getClusterName());
+      SafeHelixDataAccessor helixDataAccessor = manager.getHelixDataAccessor();
+      IdealState idealState = helixDataAccessor.getProperty(propertyKeyBuilder.idealStates(storeName));
+
+      if (idealState != null) {
+        Map<String, Map<String, String>> mapFields = idealState.getRecord().getMapFields();
+        for (Integer partitionId: storageEnginePartitionIds) {
+          if (storageEngine.isMetadataPartition(partitionId)) {
+            continue;
+          }
+          String partitionDbName = storeName + "_" + partitionId;
+          if (!mapFields.containsKey(partitionDbName)
+              || !mapFields.get(partitionDbName).containsKey(instanceHostName)) {
+            storageEngine.dropPartition(partitionId);
+          }
+        }
+        if (storageEngine.getPartitionIds().isEmpty()) {
+          removeStorageEngine(storeName);
+        }
+      } else {
+        removeStorageEngine(storeName);
+      }
+    }
```

Comment:
Can we please add some logging here to indicate that we're deleting partitions, storage engines? It would be helpful for debugging in the case we observe unnecessary boostrapping.